### ISSUE TITLE
Missing encoding data for: "MacExpertEncoding"

### ISSUE THREAD

**Issue by trapero** [smalot/pdfparser#85]
Hello,

I´m trying to parse the metadata from different files, the library works great but in some documents give me this error:

Missing encoding data for: "MacExpertEncoding"

You can access one of the document in:

http://www.statistik.rlp.de/fileadmin/dokumente/berichte/C1013_201500_1j_L.pdf

I get this error in the demo too.

Do you know a solution?

Thank you

**Comment by garraeth** [comment]
I'm getting this same message. 

A couple resources I found, but I'm no PDF dev so it's foreign to me:

https://github.com/rohitpaulk/pdfparser/blob/master/gems/pdf-reader-master/lib/pdf/reader/encodings/mac_expert.txt

https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
	page: 665

This would be a great addition.

Thanks! Love the project!

**Comment by garraeth** [comment]
Just on a whim, I created a MacExpertEncoding class (using the MacRomanEncoding.php file as a template) and copied the $encoding data from MacRomanEncoding into my new MacExpertEncoding class...and it worked!

Granted, I might have just gotten lucky.

**Comment by smalot** [comment]
Could you create a merge request to contribute ?

**Comment by adrianbj** [comment]
Sorry if I should open a new issue, but I am actually getting this error:

Missing encoding data for: ""

I have worked around it for now by specifying `StandardEncoding` like this:
```
$className = '\\Smalot\\PdfParser\\Encoding\\StandardEncoding';
```

The problem is that `$this->get('BaseEncoding')` returns `false`

Any thoughts on how to fix this properly?

Thanks.

**Comment by k00ni** [comment]
Can you provide us the PDF which is causing this problem? It must be free of charge and without any obligations.

**Comment by adrianbj** [comment]
Hi @k00ni  - thanks for responding so quickly. The issue can be seen with this file.
[healthy-chesapeake-waterways.pdf](https://github.com/smalot/pdfparser/files/5880915/healthy-chesapeake-waterways.pdf)

**Comment by k00ni** [comment]
Is `healthy-chesapeake-waterways.pdf` free of charge and without any obligations?

**Comment by adrianbj** [comment]
Yes - it's from here: https://ian.umces.edu/press/newsletters/publication/1/healthy_chesapeake_waterways_2002-05-01/

Filename is different, but that's just because I am rebuilding this site and the new version is automatically renaming PDFs to match the publication title.

**Comment by igor-krein** [comment]
Currently, in the Encoding.php, there is a "magic" function __toString() implementation, which calls a getEncodingClass() function, which, in its turn, may throw an exception (line 150).

From [php.net](https://www.php.net/manual/en/language.oop5.magic.php#object.tostring):

> **Warning** It was not possible to throw an exception from within a __toString() method prior to PHP 7.4.0. Doing so will result in a fatal error.

I think, if there is a need to keep a PHP 7.4 support, better change the line 150 somehow (for example, simply return an empty string). Note that __toString() method **must** return a string.

**Comment by suuuunto** [comment]
I am getting the same error, is there a dirty fix for it?

**Comment by k00ni** [comment]
@igor-krein and @suuuunto does #407 fix it?

**Comment by igor-krein** [comment]
> @igor-krein and @suuuunto does #407 fix it?

The code looks like it should fix the problem, thanks. I'll try to check in action asap.
